src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.erase_lead
+
+/-!
+# Reverse of a univariate polynomial
+
+The main definition is `reverse`.  Applying `reverse` to a polynomial `f : polynomial R` produces
+the polynomial with a reversed list of coefficients, equivalent to `X^f.nat_degree * f(1/X)`.
+
+The main result is that `reverse (f * g) = reverse (f) * reverse (g)`, provided the leading
+coefficients of `f` and `g` do not multiply to zero.
+-/
+
+namespace polynomial
+
+open polynomial finsupp finset
+open_locale classical
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
+namespace rev
+/-- If `i ≤ N`, then `rev_at_fun N i` returns `N - i`, otherwise it returns `i`.
+
+Essentially, this function is only used for `i ≤ N`.
+The advantage of `rev_at_fun N i` over `N - i` is that `rev_at_fun` is an involution.
+-/
+def rev_at_fun (N i : ℕ) : ℕ := ite (i ≤ N) (N-i) i
+
+@[simp] lemma rev_at_fun_invol {N i : ℕ} : rev_at_fun N (rev_at_fun N i) = i :=
+begin
+  unfold rev_at_fun,
+  split_ifs with h j,
+  { exact nat.sub_sub_self h, },
+  { exfalso,
+    apply j,
+    exact nat.sub_le N i, },
+  { refl, },
+end
+
+lemma rev_at_fun_inj {N : ℕ} : function.injective (rev_at_fun N) :=
+begin
+  intros a b hab,
+  rw [← @rev_at_fun_invol N a, hab, rev_at_fun_invol],
+end
+
+/-- `rev_at N` is `rev_at_fun N` bundled as an embedding. -/
+def rev_at (N : ℕ) : function.embedding ℕ ℕ :=
+  { to_fun := λ i , (ite (i ≤ N) (N-i) i),
+    inj' := rev_at_fun_inj }
+
```

Comment:
If we have multiple ways of writing an expression, it's a good idea to choose one as canonical. That way, we avoid having to duplicate our work for all the different ways to express the same thing. Writing out a `simp` lemma serves as kind of a Lean-readable documentation of which one we prefer.
```suggestion

/-- We prefer to use the bundled `rev_at` over unbundled `rev_at_fun`. -/
@[simp] lemma rev_at_fun_eq (N i : ℕ) : rev_at_fun N i = rev_at N i := rfl

```